server/handlers/authentication_test.go

```diff
@@ -0,0 +1,221 @@
+	responseRecorder := httptest.NewRecorder()
+	handler.Login(responseRecorder, req)
+
+	// Verify the results
+	assert.Equal(t, http.StatusSeeOther, responseRecorder.Code)
+	assert.Equal(t, "/", responseRecorder.Header().Get("Location"))
+}
+
+func Test_PostLoginHandler_ValidCredentials_OK(t *testing.T) {
+	// Prepare the test
+	var (
+		username     = "username"
+		password     = "password"
+		nextUrl      = "/next"
+		fakeJwtToken = "fakeJwt"
+
+		loginResult = map[string]interface{}{
+			"token":    fakeJwtToken,
+			"username": username,
+		}
+	)
+
+	ctrl := gomock.NewController(t)
+
+	serviceMock := mocks.NewMockServiceInterface(ctrl)
+	serviceMock.EXPECT().LoginUser(username, password).Return(loginResult, nil)
+
+	handler := NewAuthenticationHandler(serviceMock, nil)
+
+	// Execute the test
+	params := url.Values{}
+	params.Add("username", username)
+	params.Add("password", password)
+
+	req := httptest.NewRequest("POST", "/login", bytes.NewBufferString(params.Encode()))
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.URL.RawQuery = "next=" + nextUrl
+
+	responseRecorder := httptest.NewRecorder()
+	handler.Login(responseRecorder, req)
+
+	// Verify the results
+	assert.Equal(t, http.StatusSeeOther, responseRecorder.Code)
+	assert.Equal(t, nextUrl, responseRecorder.Header().Get("Location"))
+
+	var isCookieStored bool
+	for _, cookie := range responseRecorder.Result().Cookies() {
+		if cookie.Name == "token" {
+			isCookieStored = true
+			assert.Equal(t, fakeJwtToken, cookie.Value)
+		}
+	}
+
+	assert.True(t, isCookieStored)
+}
+
+func Test_PostLoginHandler_TokenNotExists_OK(t *testing.T) {
+	// Prepare the test
+	var (
+		username = "username"
+		password = "password"
+		nextUrl  = "/next"
+
+		loginResult = map[string]interface{}{
+			"username": username,
+		}
+
+		expectedLoginHTMLPath     = "assets-v1/templates/src/pages/oauth_portal/login.html"
+		expectedHTMLParsingParams = map[string]interface{}{
+			"HasNext": true,
+			"Next":    nextUrl,
+			"Error":   "could not get token",
+		}
+	)
+
+	ctrl := gomock.NewController(t)
+
+	htmlParserMock := func(w http.ResponseWriter, htmlFile string, params map[string]interface{}) {
+		assert.Equal(t, expectedLoginHTMLPath, htmlFile)
+		assert.Equal(t, expectedHTMLParsingParams, params)
+	}
+	serviceMock := mocks.NewMockServiceInterface(ctrl)
+	serviceMock.EXPECT().LoginUser(username, password).Return(loginResult, nil)
+
+	handler := NewAuthenticationHandler(serviceMock, htmlParserMock)
+
+	// Execute the test
+	params := url.Values{}
+	params.Add("username", username)
+	params.Add("password", password)
+
+	req := httptest.NewRequest("POST", "/login", bytes.NewBufferString(params.Encode()))
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.URL.RawQuery = "next=" + nextUrl
+
+	responseRecorder := httptest.NewRecorder()
+	handler.Login(responseRecorder, req)
+
+	// Verify the results
+	assert.Equal(t, http.StatusOK, responseRecorder.Code)
```

Comment:
When there is no token and request returns via "parseLoginWithError", I feel like a status code of "200" coudl be improved.

Reply:
In the subsequent test, you add this check, `assert.Len(t, responseRecorder.Result().Cookies(), 0)` and that's probably appropriate here too.

Reply:
Make sense, i will update it

Reply:
We already have the cookies assertion in the "Invalid Credentials" test case

Reply:
True. But I will make the argument that either it is unnecessary in the "Invalid Credentials" test case, or that _it should_ be included here. 

If there are invalid credentials, the function exits at line 72. If there is no token, it exits at line 79. In both cases lines 82 - 86 are NOT run. So why test in "Test_PostLoginHandler_InvalidCredentials_OK" but not "Test_PostLoginHandler_TokenNotExists_OK." In both cases, the http.SetCookie function is not run.

Reply:
eh you're right, i thought you was commenting the get handler function, let me update the code!